protobuf_serialization/writer.nim

```diff
@@ -0,0 +1,251 @@
+  sub: bool,
+  existingLength: var int
+) {.raises: [Defect, IOError, ProtobufWriteError].} =
+  if existingLength > 255:
+    raise newException(ProtobufWriteError, "Buffer length exceeded 255 when writing a new nested object.")
+  if sub:
+    existingLength += 2
+
+  var bytes: seq[byte]
+
+  #String/byte seqs.
+  when flatValue is CastableLengthDelimitedTypes:
+    if flatValue.len == 0:
+      existingLength -= 2
+      return
+    bytes = cast[seq[byte]](flatValue)
+
+  #Standard lib types which use custom converters, instead of encoding the literal Nim representation.
+  elif flatType(flatValue).isStdlib():
+    bytes = flatValue.stdlibToProtobuf()
+    if bytes.len == 0:
+      return
+
+  #Nested object which even if the sub-value is empty, should be encoded as long as it exists.
+  elif rootType.isPotentiallyNull():
+    var substream = memoryOutput()
+    writeValueInternal(substream, flatValue, true, existingLength)
+    bytes = substream.getOutput()
+
+  #Object which should only be encoded if it has data.
+  elif flatValue is object:
+    var substream = memoryOutput()
+    writeValueInternal(substream, flatValue, true, existingLength)
+    bytes = substream.getOutput()
+    existingLength -= 2
+    if bytes.len == 0:
+      return
+
+  #Distinct types.
+  else:
+    bytes = flatValue.toProtobuf()
+    if bytes.len == 0:
+      existingLength -= 2
+      return
+
+  if not sub:
+    existingLength += bytes.len
+  if existingLength > 255:
+    raise newException(ProtobufWriteError, "Buffer length exceeded 255 when appending data.")
+
+  stream.write(key(fieldNum, LengthDelimited))
+  stream.write(byte(bytes.len))
+  stream.write(bytes)
+
+proc writeFieldInternal[T](
+  stream: OutputStream,
+  fieldNum: uint,
+  value: T,
+  sub: bool,
+  existingLength: var int
+) {.raises: [Defect, IOError, ProtobufWriteError].} =
+  if false:
+    raise newException(ProtobufWriteError, "")
+
+  let flattenedOption = value.flatMap()
+  if flattenedOption.isNone():
+    return
+  let flattened = flattenedOption.get()
+
+  when flattened is bool:
+    stream.writeVarInt(fieldNum, UInt(flattened))
+  elif flattened is VarIntWrapped:
+    stream.writeVarInt(fieldNum, flattened)
+  elif flattened is FixedWrapped:
+    stream.writeFixed(fieldNum, flattened)
+  else:
+    writeLengthDelimited(stream, fieldNum, T, flattened, sub, existingLength)
+
+proc writeField*[T](
+  writer: ProtobufWriter,
+  fieldNum: uint,
+  value: T
+) {.raises: [Defect, IOError, ProtobufWriteError].} =
+  static: verifyWritable(flatType(T))
+  var existingLength = 0
+  writer.stream.writeFieldInternal(fieldNum, value, false, existingLength)
+
+proc writeValueInternal[T](
+  stream: OutputStream,
+  value: T,
+  sub: bool,
+  existingLength: var int
+) {.raises: [Defect, IOError, ProtobufWriteError].} =
+  if false:
+    raise newException(ProtobufWriteError, "")
+  if false:
+    raise newException(IOError, "")
+
+  let flattenedOption = value.flatMap()
```

Comment:
The `flatMap` approach leads to unnecessary copies though.

With the recursive approach, it would be very easy to traverse the Options and refs exactly in the same way `flaptMap` currently traverses them. In the end, the recursive process will have two possible outcomes:

1) It will terminate, because it found a `nil`/`none` value (this is similar to the `return` statement below)

2) It will reach a leaf value and the serialization will continue. The leaf value will be equivalent to the `flattened` variable below, so you can just copy paste the entire block of code and place it in the function that will be eventually called by the recursive process.